packages/react-dom-bindings/src/shared/ReactDOMFormActions.js

```diff
@@ -76,7 +76,7 @@ export function useFormStatus(): FormStatus {
 }
 
 export function useFormState<S, P>(
-  action: (S, P) => Promise<S>,
+  action: (S, P) => S | Promise<S>,
```

Comment:
In TypeScript I think this would be:

```
export function useFormState<S, P>(
  action: (Awaited<S>, P) => S,
  initialState: Awaited<S>,
  permalink?: string,
): [Awaited<S>, (P) => void] {
```

So if S is a Promise, the state is the awaited result. If S is not a promise it is the same as S.

Is there some equivalent in Flow?

Reply:
there’s not one built in but I asked the Flow team about it and they showed me how to create an equivalent, just forgot to update the PR

Reply:
I can't speak for Flow, but in TypeScript the `action` annotation as chosen by Andrew is the idiomatic way of declaring a callback that can be sync or async, if I'm not mistaken. The usage of `Awaited` as proposed here would actually lead to a compiler error for async actions, as demonstrated in [this TypeScript playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtX2QGcQAxHGAWwGUMoMQBGAHioBp4AFAPgAoAoePCiZcqAFzwehWvQlt4ABxg4FhCRwCU8ALxd4VeAB9OyilmIsurAfCyos2KBBp0Qc64IUhKTuwGsAfglpGDsAc2sNCQBteR4lFTVOLV14ADccLGAAXQBuPj5QSFgEFHRsPAJiMkoXegAmFnZufkFhCvFJaVcJAEEAdygHEGBLdgTVdRS9Nhs7BywnOrd4AaH6UaorGy8fCH8g+BDwyJi14c2rSQmkzR09DKy8gqJScmoZRh4pT+mjz-YAAYNPlXjUPq4GDwoIQAJ7oLq-e7-VxAkF8MHvZb1b7deh-PEgNGg6pYz44mHwsCI1wEgHwYG5IA).

Reply:
If your intention with using `Awaited` is to disallow a promise being used as initial state, which the annotation did not prevent before but maybe should, a possible solution would be to use a combination of both proposed signatures:

```ts
function useFormState<S, P>(
  action: (prevState: Awaited<S>, payload: P) => S | Promise<S>,
  initialState: Awaited<S>,
  permalink?: string,
): [Awaited<S>, (payload: P) => void];
```

See also [this playground](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtX2QGcQAxHGAWwGUMoMQBGAHioBp4AFAPgAoAoePCiZcqAFzweABzgA3GnRAS28KVACeEHFGASOASngBeLvCrwAPpxg4KWYiy6sB8LKizYoEBfWXPBUiCUXm4A1gD8EoQYMG4A5s76EgDaKtIaWjp6hibwsjhYwAC6ANx8fKCQsAgo6Nh4BMRklD4gAEws7Nz8gsL14pIyIPK0vvAAggDuUB4gwI7sapraupw5pmwubh5YXq0SUzP081ROLoHBEGGR8NGxqAl8SfDJh7MnToMZK9nGpvmFUrlSrQOBINAiBpEUjkaijEAAZk6nF4Lj6ogk0jk+wm03eC1U3yyaz+Zks1ls9hACy27k83nhBzxxxpASCFBCqAiURi8USKTeLNO7HSy2JBlJAOKZT40OacMUDB4WOGrXWqmx8PYAAZ9GU5bDWkqoIR1OhBprFOqhiNFDq9XwAPSO+AAAQwhAAtCAAB6BTDemA2GDwAAqAAt7PBJjhkBBgPAICAdPAMDhU+GajgIFpJvF4DA0NgKAggsGxE6XSazWBHVBma5CPBUDgMPB1CA24RkFIpORjq58ABhK4gVBtoe2PuoMce9h4CDqMxBWRBeCTijT2eEAB0lbDkabZfIjfgOEQ9HwYCg0ITACMl8IwIVZ15F0JgMB8wADADk0PgSAsFnX9v1TdMoHgCgcGAOMQH3DBwzoaMTTPWI4jcN8l0mWIMEvJAT0QhBiBgVcYD3A0WnhY1TXNHhoitUkGPoLobDsYgdzgQhs1XHhdQdWUmkNeE2mVG01VJcStXgXV9SEqjFFE6s6KkxjclUliZIdZ03Q9b0-XADBA2DA8oxjOMEyTFM0wzLMcxwPMHgLIssBLeBjxgCsdOU2t6yOU8WzbDsux7PsYAHNx11Hcd1ynPBt3nVB3yoFc1w3Ldx13fcIyjDzT3PfDr1veAHyEMBn1AccsI-L8nL-ACgJAsCbMg6DYKTBCkLbaYm3ILAMNQaqcI8fDEEIzNblS8jBJhBT6CU2iwEkZiQGtS1NI4NiqU4kBuIgXj+JlSiFXoBExPW1bJIu+05Nmk7ER4HyLVVeE1peu0tLKHT3S9X1-SMvKcqbcz40TZMExsoiCPsxy4mc8dXNLINyC8qtFrrBso0C9tO1uUL+zmQcouAmL0vizLEuSqbYs3cmPT3HSgfc5GQyjAqx0Am9iHvR9ypfKqc0fT8f3-YhAOijBQPAoQoJguDOuQnq0P6zDBejXDRvG4ipoo+T7rOp6VVteg3uNkBWMpDiuJ4kA+P0B0gA) with a comparison of all three solutions.

Reply:
Huh, I don't really understand why the `S | Promise<S>` thing is necessary. I'm just going to let @eps1lon figure this out in the DefinitelyTyped repo since that's the one that actually matters 😆 For our internal Flow types, I updated it to @sebmarkbage 's version (https://github.com/facebook/react/pull/27571/commits/165424e8b31bf6490ebac48708d5231dff701b1a), which mostly worked but I did hit a few "recursion limit exceeded" errors and ended up needing to resort to `any` in a few places anyway. But as long as the public ones are correct it's not that important.